apps/emqx_retainer/src/emqx_retainer_dispatcher.erl

```diff
@@ -334,25 +334,62 @@ deliver_in_batches(Msgs, BatchSize, Pid, Topic, Limiter0) ->
             Drop
     end.
 
+-define(DELIVER_ALLOWED, true).
+-define(DELIVER_NOT_ALLOWED, false).
+
 deliver_to_client([Msg | T], Pid, Topic) ->
+    Pipe = emqx_utils:pipeline(
+        [
+            fun check_clientid_banned/2,
+            fun 'check_prefixed_$_with_wildcard'/2
+        ],
+        {Msg, Topic},
+        ?DELIVER_NOT_ALLOWED
+    ),
     _ =
-        case emqx_banned:check_clientid(Msg#message.from) of
-            false ->
+        case Pipe of
+            {ok, _, ?DELIVER_ALLOWED} ->
                 Pid ! {deliver, Topic, Msg};
-            true ->
-                ?tp(
-                    notice,
-                    ignore_retained_message_deliver,
-                    #{
-                        reason => "client is banned",
-                        clientid => Msg#message.from
-                    }
-                )
+            {error, _, _} ->
+                ok
         end,
     deliver_to_client(T, Pid, Topic);
 deliver_to_client([], _, _) ->
     ok.
 
+check_clientid_banned({Msg, _Topic} = Input, _) ->
+    case emqx_banned:check_clientid(Msg#message.from) of
+        false ->
+            {ok, Input, ?DELIVER_ALLOWED};
+        true ->
+            ?tp(
+                notice,
+                ignore_retained_message_deliver,
+                #{
+                    reason => client_is_banned,
+                    clientid => Msg#message.from
+                }
+            ),
+            {error, client_is_banned, ?DELIVER_NOT_ALLOWED}
+    end.
+
+'check_prefixed_$_with_wildcard'({Msg, Topic} = Input, _) ->
+    case emqx_topic:match(Msg#message.topic, Topic) of
+        false ->
+            ?tp(
+                notice,
+                ignore_retained_message_deliver,
```

Comment:
Nit. Doesn't seem like a good idea to me to log `notice` event in this case to be honest. After all, it's the same as unmatched topic filter, which is not even a log event right now, rightly so.

Reply:
`Msg` here are absolutely matched to a given `Topic`
(Use topic index to retrieve from retainer table)

The topic matching is actually to ensure that messages starting with `$` are not sent to subscriptions starting with `+` or wildcard sub `#`

> The Server MUST NOT match Topic Filters starting with a wildcard character (# or +) with Topic Names beginning with a $ character [MQTT-4.7.2-1]

So in general, the retain messages to be delivered are filtered here, and the log level should be `notice` (I just temporarily used the same notice as the client banned check) or `info`.
They should all use the `info` level to me.

Reply:
Nit:
Message filtering should be performed before attempting to deliver any messages

```diff
 deliver(Messages, Pid, Topic, Limiter) ->
     case erlang:is_process_alive(Pid) of
         false ->
             no_receiver;
         _ ->
             BatchSize = emqx_conf:get([retainer, flow_control, batch_deliver_number], undefined),
+            NMessages = filter_delivery(Messages, Topic),
             case BatchSize of
                 0 ->
-                    deliver_to_client(Messages, Pid, Topic),
+                    deliver_to_client(NMessages, Pid, Topic),
                     {ok, Limiter};
                 _ ->
-                    deliver_in_batches(Messages, BatchSize, Pid, Topic, Limiter)
+                    deliver_in_batches(NMessages, BatchSize, Pid, Topic, Limiter)
             end
     end.
```

Reply:
>  They should all use the info level to me.

We're just following the spec, IMO there's no point in logging anything other than `debug` event, or perhaps just a tracepoint.

Reply:
right, debug is enough when following spec 😎